plugin howto bootstrap (#1213)
pulp/pulp-cli: docs/dev/guides/bootstrap.md

- `pyproject.toml`: Project metadata, dependencies, and build configuration. This file contains essential project information including author details, licensing information, version numbers, package dependencies, and build system configuration. You'll need to customize the author name, email, project description, license type to match your plugin's specific requirements.
- `CHANGES/`: Directory for changelog fragments (.feature, .bugfix, etc. files). See the [changelog update guide](https://pulpproject.org/pulpcore/docs/dev/guides/git/?h=towncrier#changelog-update) for more details.
- `.github/workflows/`: This directory contains GitHub Actions workflow configurations for automated testing and continuous integration. These files are pre-configured and do not require modification when initially setting up your plugin development environment. The CI workflows will automatically run tests and validation checks on your plugin code when you push changes to your repository.

---

## Customizing Your Plugin

### Create API Context Classes

Edit `pulp-glue-my-plugin/pulp_glue/my_plugin/context.py` to define context classes that handle API interactions:

```python
import typing as t

> from gettext import gettext as _

Reviewer: This should be using the `from pulp_glue.common.i18n import get_translation` but again we are not even using it in this example.
```suggestion
```

> from pulp_glue.common.context import PulpContext, PulpEntityContext

Reviewer: ```suggestion
from pulp_glue.common.context import PulpEntityContext
```

The linter would complain later, but atm we are not using it.

class PulpMyExampleContext():
    """Context for working with my custom resource."""

>     ID_PREFIX = "my_resource"

Reviewer: Can you add a `NEEDS_PLUGINS` here too?

    def example_action(self, data: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
        """Create a custom resource with specific data.

        Args:
            data: The data dictionary to send to the API

        Returns:
            The created resource entity
        """
        response = self.call(

>             operation_id="my_resource_example_action",

Reviewer: ```suggestion
            "example_action",
```

            body=data,

>             validate_body=False,

Reviewer: ```suggestion
```
That's meant to be a last resort. Whenever we can we want to encourage linting on the client side before bothering the server.

# Import your command plugins
from pulpcore.cli.my_plugin.my_command import my_command_group

translation = get_translation(__package__)
_ = translation.gettext

__version__ = "0.1.0"  # Matches version in pyproject.toml

@pulp_group(name="my_plugin")
def my_plugin_group() -> None:
    """My plugin commands."""
    pass

def mount(main: click.Group, **kwargs: t.Any) -> None:
    # Add your command groups here
    my_plugin_group.add_command(my_command_group)
    main.add_command(my_plugin_group)
```

### Create Command plugins

Create a new file for your commands, e.g., `pulpcore/cli/my_plugin/my_command.py`:

```python
import click
from pulp_cli.common.context import pass_pulp_context
from pulp_glue.common.context import PulpContext
from pulp_glue.my_plugin.context import PulpMyExampleContext
from pulp_cli.common.generic import pass_entity_context

@click.group()
@pass_pulp_context
@click.pass_context 
def my_command_group(ctx: click.Context, pulp_ctx: PulpContext, /) -> None:
    """My custom commands."""
    ctx.obj = PulpMyExampleContext(pulp_ctx)

@my_command_group.command()
@click.option("--data", required=True, help="Data for the example action")

> @pass_pulp_context
> @pass_entity_context
> def create(pulp_ctx: PulpContext, data: str):
>     """Create a new item using example action."""
>     my_example_ctx = PulpMyExampleContext(pulp_ctx)

Reviewer: ```suggestion
@pass_entity_context
@pass_pulp_context
def create(pulp_ctx: PulpContext, my_example_ctx: PulpMyExampleContext, /, data: str):
    """Calling example action."""
```

from pulpcore.cli.my_plugin.my_command import my_command_group

translation = get_translation(__package__)
_ = translation.gettext

__version__ = "0.1.0"  # Matches version in pyproject.toml

@pulp_group(name="my_plugin")
def my_plugin_group() -> None:
    """My plugin commands."""
    pass

def mount(main: click.Group, **kwargs: t.Any) -> None:
    # Add your command groups here
    my_plugin_group.add_command(my_command_group)
    main.add_command(my_plugin_group)
```

### Create Command plugins

Create a new file for your commands, e.g., `pulpcore/cli/my_plugin/my_command.py`:

```python
import click
from pulp_cli.common.context import pass_pulp_context
from pulp_glue.common.context import PulpContext
from pulp_glue.my_plugin.context import PulpMyExampleContext
from pulp_cli.common.generic import pass_entity_context

@click.group()
@pass_pulp_context
@click.pass_context 
def my_command_group(ctx: click.Context, pulp_ctx: PulpContext, /) -> None:
    """My custom commands."""
    ctx.obj = PulpMyExampleContext(pulp_ctx)

@my_command_group.command()
@click.option("--data", required=True, help="Data for the example action")
@pass_entity_context
@pass_pulp_context

> def create(pulp_ctx: PulpContext, my_example_ctx: PulpMyExampleContext, /, data: str):

Reviewer: We shouldn't call this create anymore...

    """Calling example action."""
    result = my_example_ctx.example_action({"data": data})
    pulp_ctx.output_result(result)
```

---

## Development Workflow

### Install Your plugin in Development Mode

For installation instructions, see here [installation.md].

### Test Your Commands

After installation, you can test your commands:

```bash
pulp my-plugin my-command create --data "example"
```

Consider writing tests for your commands as soon as you implement them.
[testing doc]

---

## Update Templates

If you need to update the project structure with new template changes:

```bash
cd pulp-cli-my-plugin
../pulp-cli/cookiecutter/apply_templates.py
```

---

> By using the bootstrap process, you can set up a properly structured Pulp CLI plugin and focus on implementing your custom commands rather than worrying about project structure.

Reviewer: I wonder what the intention of this paragraph is. When someone made it to here, I hope they knew what they'd been doing.
It sounds more like a motivational thing that can go in the frontmatter. But there is something similar too.
Can we merge them somehow?
Or as being a guide it should read like " At this point you have accomplished ..., next steps are...".

# How to Create a Pulp CLI Plugin

This guide shows you how to create a Pulp CLI plugin using the provided bootstrap templates.
Instead of manually creating all the required files, the bootstrap process will automatically generate the correct structure and files for you.

> ---
> 

Reviewer: ```suggestion
```

## Overview

Pulp CLI plugin extend the functionality of the Pulp CLI by adding custom commands and features.
The pulp-cli project includes templates that make it easy to bootstrap a new plugin with the correct structure and configuration.

> 
> ---

Reviewer: ```suggestion
```

## Prerequisites

Before you begin, ensure you have the following installed:

```bash
pip install cookiecutter click pyyaml tomlkit
```

> 
> ---

Reviewer: ```suggestion
```

## Bootstrap a New Pulp CLI Plugin

### Using the Bootstrap Template

Navigate to the directory where you want to create your new plugin and run:

```bash

> # Clone the pulp-cli repository if you don't have it already
> git clone https://github.com/pulp/pulp-cli.git
> 
> # Run the bootstrap script
> pulp-cli/cookiecutter/apply_templates.py --bootstrap
> 

Reviewer: ```suggestion
# Clone the pulp-cli repository if you don't have it already
git clone https://github.com/pulp/pulp-cli.git

# Run the bootstrap script
pulp-cli/cookiecutter/apply_templates.py --bootstrap
```

```

### Answer the Prompts

You will be prompted for several values during the bootstrap process.
The most important is **app_label**, which by convention (when applicable) should match the Pulp server component your plugin targets (for example, use `file` for `pulp-file`).
This value determines the import paths and command names for your plugin.

> You will also see additional prompts for options such as glue integration, documentation, translations, version, repository URL, and CI configuration—answer these as appropriate for your project.

Reviewer: ```suggestion
You will also see additional prompts for options such as glue integration, documentation, translations, version, repository URL, and CI configuration.
Answer these as appropriate for your project.
```

### Key Files and Directories

The bootstrap process creates several important directories and files:

> - `pulpcore/cli/my-plugin/__init__.py`: Main entry point where you define command groups.
> See bellow for more details
> - `pulp-glue-my-plugin/pulp_glue/my-plugin/context.py`: API context class for interacting with Pulp.
> See bellow for more details

Reviewer: ```suggestion
- `pulpcore/cli/my-plugin/__init__.py`: Main entry point where you define command groups.
  See below for more details
- `pulp-glue-my-plugin/pulp_glue/my-plugin/context.py`: API context class for interacting with Pulp.
  See below for more details
```

> - `pyproject.toml`: Project metadata, dependencies, and build configuration.
> This file contains essential project information including author details, licensing information, version numbers, package dependencies, and build system configuration.
> You'll need to customize the author name, email, project description, license type to match your plugin's specific requirements.
> - `CHANGES/`: Directory for changelog fragments (.feature, .bugfix, etc. files).
> See the [changelog update guide] for more details.
> - `.github/workflows/`: This directory contains GitHub Actions workflow configurations for automated testing and continuous integration.
> These files are pre-configured and do not require modification when initially setting up your plugin development environment.
> The CI workflows will automatically run tests and validation checks on your plugin code when you push changes to your repository.

Reviewer: ```suggestion
- `pyproject.toml`: Project metadata, dependencies, and build configuration.
  This file contains essential project information including author details, licensing information, version numbers, package dependencies, and build system configuration.
  You'll need to customize the author name, email, project description, license type to match your plugin's specific requirements.
- `CHANGES/`: Directory for changelog fragments (.feature, .bugfix, etc. files).
  See the [changelog update guide] for more details.
- `.github/workflows/`: This directory contains GitHub Actions workflow configurations for automated testing and continuous integration.
  These files are pre-configured and do not require modification when initially setting up your plugin development environment.
  The CI workflows will automatically run tests and validation checks on your plugin code when you push changes to your repository.
```

> 
> ---

Reviewer: ```suggestion
```

## Customizing Your Plugin

### Create API Context Classes

Edit `pulp-glue-my-plugin/pulp_glue/my_plugin/context.py` to define context classes that handle API interactions:

```python
import typing as t

from pulp_glue.common.context import (
    PulpEntityContext,
    PluginRequirement
) 

> class PulpMyExampleContext():

Reviewer: By convention we should call this
```suggestion
class PulpMyResourceContext():
```

    """Context for working with my custom resource."""

    ID_PREFIX = "my_resource"
    NEEDS_PLUGINS = [PluginRequirement("my_plugin", specifier=">=0.0.1")]

    def example_action(self, data: t.Dict[str, t.Any]) -> t.Dict[str, t.Any]:
        """Create a custom resource with specific data.

        Args:
            data: The data dictionary to send to the API

        Returns:
            The created resource entity

>         """

Reviewer: This is not matching the example_action.

>         response = self.call(

Reviewer: ```suggestion
        # TODO you probably need to compile `data` from more detailed options here.
        response = self.call(
```

            operation="example_action",
            body=data,
        )
        return response
```

> ### Modify Command Group

Reviewer: Please move this section at the end of the example.
Also let's describe it as:
```suggestion
### Add Subcommands to the CLI
```

Edit `pulpcore/cli/my_plugin/__init__.py` to define your command groups and add functionality:

```python
import typing as t
import click
from pulp_cli.generic import pulp_group
from pulp_glue.common.i18n import get_translation

# Import your command plugins
from pulpcore.cli.my_plugin.my_command import my_command_group

translation = get_translation(__package__)
_ = translation.gettext

__version__ = "0.1.0"  # Matches version in pyproject.toml

@pulp_group(name="my_plugin")
def my_plugin_group() -> None:
    """My plugin commands."""
    pass

def mount(main: click.Group, **kwargs: t.Any) -> None:
    # Add your command groups here
    my_plugin_group.add_command(my_command_group)
    main.add_command(my_plugin_group)
```

> ### Create Command plugins

Reviewer: ```suggestion
### Create Plugin Commands
```

> Create a new file for your commands, e.g., `pulpcore/cli/my_plugin/my_command.py`:

Reviewer: ```suggestion
Create a new file for your resource, e.g., `pulpcore/cli/my_plugin/my_resource.py`:
```
Please adjust the import too.

```python
import click
from pulp_cli.common.context import pass_pulp_context
from pulp_glue.common.context import PulpContext

> from pulp_glue.my_plugin.context import PulpMyExampleContext

Reviewer: ```suggestion
from pulp_glue.my_plugin.context import PulpMyResourceContext
```

from pulp_cli.common.generic import pass_entity_context

@click.group()
@pass_pulp_context
@click.pass_context 

> def my_command_group(ctx: click.Context, pulp_ctx: PulpContext, /) -> None:

Reviewer: ```suggestion
def my_resource(ctx: click.Context, pulp_ctx: PulpContext, /) -> None:
```

    """My custom commands."""

>     ctx.obj = PulpMyExampleContext(pulp_ctx)

Reviewer: ```suggestion
    ctx.obj = PulpMyResourceContext(pulp_ctx)
```

> @my_command_group.command()

Reviewer: ```suggestion
@my_resource.command()
```

@click.option("--data", required=True, help="Data for the example action")
@pass_entity_context
@pass_pulp_context
def my_command(pulp_ctx: PulpContext, my_example_ctx: PulpMyExampleContext, /, data: str):
    """Calling example action."""

>     result = my_example_ctx.example_action({"data": data})

Reviewer: ```suggestion
    result = my_resource_ctx.example_action({"data": data})
```

    pulp_ctx.output_result(result)
```

> 
> ---

Reviewer: ```suggestion
```

## Development Workflow

### Install Your plugin in Development Mode

For installation instructions, see here [installation.md].

### Test Your Commands

After installation, you can test your commands:

```bash

> pulp my-plugin my_command_group my_command --data "example"

Reviewer: ```suggestion
pulp my-plugin my-resource my-command --data "example"
```

```

> Consider writing tests for your commands as soon as you implement them.
> [testing doc]

Reviewer: ```suggestion
Consider [writing tests] for your commands as soon as you implement them.
```

> 
> ---

Reviewer: ```suggestion
```

## Update Templates

If you need to update the project structure with new template changes:

```bash
cd pulp-cli-my-plugin
../pulp-cli/cookiecutter/apply_templates.py
```

> 
> ---
> 

Reviewer: ```suggestion
```

[changelog update guide]: https://pulpproject.org/pulpcore/docs/dev/guides/git/?h=towncrier#changelog-update
[installation.md]: https://github.com/pulp/pulp-cli/blob/main/docs/user/guides/installation.md#from-a-source-checkout

> [testing doc]: https://pulpproject.org/pulp-cli/docs/dev/guides/contributing/#testing

Reviewer: ```suggestion
[writing tests]: https://pulpproject.org/pulp-cli/docs/dev/guides/contributing/#testing
```
